Lib/fontTools/colorLib/builder.py

```diff
@@ -59,7 +63,96 @@
 ]
 _AffineInput = Union[_AffineTuple, ot.Affine2x3]
 
+
 MAX_PAINT_COLR_LAYER_COUNT = 255
+_DEFAULT_ALPHA = VariableFloat(1.0)
+
+# TODO: where does this go? - good comment, don't want to lose
+# COLRv1 Affine2x3 uses the same column-major order to serialize a 2D
+# Affine Transformation as the one used by fontTools.misc.transform.
+# However, for historical reasons, the labels 'xy' and 'yx' are swapped.
+# Their fundamental meaning is the same though.
+# COLRv1 Affine2x3 follows the names found in FreeType and Cairo.
+# In all case, the second element in the 6-tuple correspond to the
+# y-part of the x basis vector, and the third to the x-part of the y
+# basis vector.
+# See https://github.com/googlefonts/colr-gradients-spec/pull/85
+
+
+def _afterBuildPaintLinearGradient(paint):
+    if not hasattr(paint, "x2"):
+        paint.x2 = paint.x1
+    if not hasattr(paint, "y2"):
+        paint.y2 = paint.y1
+    return paint
+
+
+def _beforeBuildPaintRadialGradient(paint, source):
+    # normalize input types (which may or may not specify a varIdx)
+    x0 = convertTupleClass(VariableInt, source.get("x0", 0.0))
+    y0 = convertTupleClass(VariableInt, source.get("y0", 0.0))
+    r0 = convertTupleClass(VariableInt, source.get("r0", 0.0))
+    x1 = convertTupleClass(VariableInt, source.get("x1", 0.0))
+    y1 = convertTupleClass(VariableInt, source.get("y1", 0.0))
+    r1 = convertTupleClass(VariableInt, source.get("r1", 0.0))
+
+    # TODO apparently no builder_test confirms this works (?)
+
+    # avoid abrupt change after rounding when c0 is near c1's perimeter
+    c = round_start_circle_stable_containment(
+        (x0.value, y0.value), r0.value, (x1.value, y1.value), r1.value
+    )
+    x0, y0 = x0._replace(value=c.centre[0]), y0._replace(value=c.centre[1])
+    r0 = r0._replace(value=c.radius)
+
+    # update source to ensure paint is built with corrected values
+    source["x0"] = x0
+    source["y0"] = y0
+    source["r0"] = r0
+    source["x1"] = x1
+    source["y1"] = y1
+    source["r1"] = r1
+
+    return paint, source
+
+
+def _defaultColorIndex():
+    colorIndex = ot.ColorIndex()
+    colorIndex.Alpha = _DEFAULT_ALPHA
+    return colorIndex
+
+
+def _defaultColorLine():
+    colorLine = ot.ColorLine()
+    colorLine.Extend = ExtendMode.PAD
+    return colorLine
+
+
+def _buildPaintCallbacks():
+    return {
+        (
+            BuildCallback.AFTER_BUILD,
+            ot.Paint,
+            ot.PaintFormat.PaintLinearGradient,
+        ): _afterBuildPaintLinearGradient,
+        (
+            BuildCallback.BEFORE_BUILD,
+            ot.Paint,
+            ot.PaintFormat.PaintRadialGradient,
+        ): _beforeBuildPaintRadialGradient,
+        # Shouldn't default info be in otData?
+        (BuildCallback.CREATE_DEFAULT, ot.ColorIndex): _defaultColorIndex,
+        (BuildCallback.CREATE_DEFAULT, ot.ColorLine): _defaultColorLine,
+    }
+
+
+def _build(cls, source):
```

Comment:
Yes, I didn't think clients would generally want to use these

Reply:
Hm. Upon further thought I made buildPaint public and left _build private.

Reply:
Specifically, a client was assumed to want to use LayerV1ListBuilder, which does have a public buildPaint

Reply:
it looks like these `_build` and `_buildPaint` are only ever used in the builder_test.py, maybe you can move them to the test module itself? They seem to me nothing more than syntactic sugar.

Reply:
also, if the public API is deemed to be LayerV1ListBuilder.buildPaint maybe the test themselves should only ever use that method, instead of this static _buildPaint

Reply:
+1 to using the public api, will fix